geomesa-accumulo/geomesa-accumulo-datastore/src/main/scala/org/locationtech/geomesa/accumulo/iterators/KryoLazyAgeOffFilter.scala

```diff
@@ -0,0 +1,86 @@
+/***********************************************************************
+ * Copyright (c) 2013-2017 Commonwealth Computer Research, Inc.
+ * All rights reserved. This program and the accompanying materials
+ * are made available under the terms of the Apache License, Version 2.0
+ * which accompanies this distribution and is available at
+ * http://www.opensource.org/licenses/apache2.0.php.
+ ***********************************************************************/
+
+package org.locationtech.geomesa.accumulo.iterators
+
+import org.apache.accumulo.core.data.{Key, Value}
+import org.apache.accumulo.core.iterators.{Filter, IteratorEnvironment, SortedKeyValueIterator}
+import org.apache.hadoop.io.Text
+import org.locationtech.geomesa.accumulo.iterators.KryoLazyAgeOffFilter.Options
+import org.locationtech.geomesa.features.SerializationOption.SerializationOptions
+import org.locationtech.geomesa.features.kryo.{KryoBufferSimpleFeature, KryoFeatureSerializer}
+import org.locationtech.geomesa.index.iterators.IteratorCache
+import org.locationtech.geomesa.utils.geotools.RichSimpleFeatureType.RichSimpleFeatureType
+import org.opengis.feature.simple.{SimpleFeature, SimpleFeatureType}
+
+/**
+  * Abstract base implementation of an iterator that ages off data based on some strategy in a SimpleFeature
+  * stored in GeoMesa. Clients can implement logic to determine whether or not a SimpleFeature should be kept
+  * or not based on the accept method. SimpleFeatures not "accepted" will be aged off.
+  *
+  * Concrete implementations of this iterator can be configured on scan, minc, and majc to age off data
+  *
+  * Age off iterators can be stacked but this may have performance implications
+  */
+abstract class KryoLazyAgeOffFilter extends Filter {
+
+  protected var spec: String = _
+  protected var sft: SimpleFeatureType = _
+  protected var kryo: KryoFeatureSerializer = _
+  protected var reusableSF: KryoBufferSimpleFeature = _
+  protected var reuseText: Text = _
+
+  override def deepCopy(env: IteratorEnvironment): SortedKeyValueIterator[Key, Value] = {
+    val copy = super[Filter].deepCopy(env).asInstanceOf[KryoLazyAgeOffFilter]
+    copy.spec = spec
+    copy.sft = sft
+
+    val kryoOptions = if (copy.sft.getSchemaVersion < 9) SerializationOptions.none else SerializationOptions.withoutId
```

Comment:
might want to update this to take the index type (although that will break compatibility with 1.2.x): see https://github.com/locationtech/geomesa/blob/geomesa_2.11-1.3.1/geomesa-accumulo/geomesa-accumulo-datastore/src/main/scala/org/locationtech/geomesa/accumulo/iterators/KryoLazyFilterTransformIterator.scala#L59-L62